src/fs/types.rs

```diff
@@ -0,0 +1,211 @@
+use std::{hash::{Hash, Hasher}, path::{Path, PathBuf}};
+use std::ops::Deref;
+
+use serde::{Serialize, Deserialize};
+use smash_arc::{Hash40, Region};
+
+#[derive(Debug)]
+pub enum RejectionReason {
+    DuplicateFile(PathBuf),
+    NotFound(SmashPath),
+    MissingExtension
+}
+
+#[repr(transparent)]
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+pub struct SmashPath(pub PathBuf);
+
+impl Deref for SmashPath {
+    type Target = Path;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl AsRef<Path> for SmashPath {
+    fn as_ref(&self) -> &Path {
+        &self
+    }
+}
+
+impl From<SmashPath> for PathBuf {
+    fn from(smashpath: SmashPath) -> Self {
+        smashpath.0
+    }
+}
+
+impl SmashPath {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn from<P: AsRef<Path>>(path: P) -> Result<Self, String> {
+        let path = path.as_ref();
+
+        if path.extension().is_some() {
+            Ok(Self(path.to_path_buf()))
+        } else {
+            Err(String::from("This path does not have an extension"))
+        }
+    }
+
+    pub fn as_path(&self) -> &Path {
+        &self
+    }
+
+    pub fn hash40(&self) -> Result<Hash40, String> {
+        let smash_path = self.to_smash_path();
+
+        smash_path.to_str()
+            .map(Hash40::from)
+            .ok_or_else(|| format!("Couldn't convert {} to an &str", self.as_path().display()))
```

Comment:
Could this be replaced by something like...

```rust
struct InvalidUtf8PathError;

impl fmt::Display for InvalidUtf8PathError {
    fn fmt(f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Couldn't convert path to an &str")
    }
}
```
and then change hash40's return type to `Result<Hash40, InvalidUtf8PathError>` ? this would likely make the return value more efficient and a bit cleaner